### ISSUE TITLE
Poor performance at scale

### ISSUE THREAD

**Issue by bzaks1424** [insightindustry/sqlathanor#76]
Hey - I hate to be Mr. Ticket fixit - but I'm getting pretty rough performance numbers when serializing to dict

Query Time = 0:00:00.199226
Serialization Time = 0:00:14.569234
Home Grown Serialization Time = 0:00:00.104587

I'll submit my encoder code here shortly in a few hours.

**Comment by bzaks1424** [comment]
So this is my "serialize_to_dict" function - just called "encoder"

```
def encoder(ThingObj, EncodeRelationship=True, Debug=False):
    thingType = type(ThingObj)
    fields = {}
    ###########################################################################
    inspObj = inspect(ThingObj)
    for field in inspObj.mapper.all_orm_descriptors:
        prObj = None
        propName = None
        #######################################################################
        if(field.extension_type.name == 'NOT_EXTENSION'):
            if(hasattr(field, 'key') and not field.key.startswith('_')):
                prObj = getattr(ThingObj, field.key)
                propName = field.key
        elif(field.extension_type.name == 'ASSOCIATION_PROXY'):
            prObj = getattr(ThingObj, field.value_attr)
            propName = field.value_attr
        else:
            raise Exception("Can't Handle: " + field.extension_type.name)
        if(propName is None or propName == "phash"):
            continue
        #######################################################################
        # prObjType = type(prObj)
        # if(Debug):
        #     print(propName + " is " + str(prObjType))
        if(
            isinstance(prObj, (int, float, bool, str))
            or prObj is None
        ):
            fields[propName] = prObj
        elif(isinstance(prObj, datetime.date)):
            fields[propName] = prObj.isoformat()
        elif(isinstance(prObj, list)):
            if(EncodeRelationship):
                fields[propName] = []
                for secObj in prObj:
                    fields[propName].append(encoder(secObj, Debug=Debug))
        else:
            if(propName not in inspect(thingType).relationships):
                fields[propName] = str(prObj)
            else:
                if(EncodeRelationship):
                    fields[propName] = encoder(prObj, Debug=Debug)
    return fields
```

**Comment by insightindustry** [comment]
Hmm. These are some interesting (and concerning!) findings. I have some questions:

1. Am I understanding correctly that when serializing your objects to `dict` you're seeing 14 second (!!!) serialization times per instance on average? That's rather shocking, and not at all the kind of performance that I've seen at scale in numerous real-world examples, so I'd love to figure out what's driving this (terrible) performance.

2. Can you share with me the structure of your model and its serialization configuration? That'll be helpful for me to test the serialization performance and see if I can replicate the bad performance numbers that you're seeing (or at least get close to the bad performance).

3. Am I correct that your `encoder()` function which you shared is your "home-grown" solution that is returning significantly better performance? (obviously, the home-grown encoder has an entirely different functional scope and is version-specific to SQLAlchemy 1.3+, Python 3.7+, etc. so it's not surprising that it'll have better performance since it has to handle a much narrower scope - but still, the difference in performance that you're seeing is staggering, if I understand your benchmarks correctly).

4. Can you share with me the output of your `pip list` in the environment where this is running? I'll want to mirror the environment as closely as possible to better see if I can replicate these benchmark results.

**Comment by bzaks1424** [comment]
Yes I can to all of the above - but I'd appreciate not doing this on a public forum. Any chance we can DM each other somehow? (You can find me on Twitter)

**Comment by insightindustry** [comment]
Thanks! I really appreciate the help. I’m not really active on Twitter much these days, but you can DM me @InsightIndustry there, or email me at software at insightindustry.com (space and word at is to fool spambots).

**Comment by wojtek-viirtue** [comment]
Shopping for a serialization / deserialization library and came across this.  Is this still an issue?  Was it in fact a legitimate issue?

**Comment by bzaks1424** [comment]
Hey @wojtek-viirtue - 

Just so you know - this only seemed to be an issue when I was using it in combination with the Chalice framework by AWS (http://chalice.readthedocs.io/)

So you should be good to go from your perspective.

**Comment by insightindustry** [comment]
@wojtek-viirtue I haven't been able to reproduce this issue yet, so it seems to be coupled to the use of the Chalice framework as @bzaks1424 mentioned.

**Comment by digitalkaoz** [comment]
@insightindustry i've seen the same performance bottle neck and it seems to be fixed by #97 (at least on my setup i got a x8 speed improvement)
